src/server/electron.ts

```diff
@@ -0,0 +1,207 @@
+    this.emit(ElectronEvents.ElectronApplication.Window, page);
+  }
+
+  windows(): Page[] {
+    return [...this._windows];
+  }
+
+  async newBrowserWindow(options: any): Promise<Page> {
+    const windowId = await this.evaluate(async ({ BrowserWindow }, options) => {
+      const win = new BrowserWindow(options);
+      win.loadURL('about:blank');
+      return win.id;
+    }, options);
+
+    for (const page of this._windows) {
+      if ((page as any)._browserWindowId === windowId)
+        return page;
+    }
+
+    return new Promise(resolve => {
+      const handler = (page: Page) => {
+        if ((page as any)._browserWindowId === windowId) {
+          this.off(ElectronEvents.ElectronApplication.Window, handler);
+          resolve(page);
+        }
+      };
+      this.on(ElectronEvents.ElectronApplication.Window, handler);
+    });
+  }
+
+  context(): BrowserContext {
+    return this._browserContext;
+  }
+
+  async close() {
+    await this.evaluate(({ app }) => app.quit());
+    this._nodeConnection.close();
+  }
+
+  async _init()  {
+    this._nodeSession.once('Runtime.executionContextCreated', event => {
+      this._nodeExecutionContext = new js.ExecutionContext(new CRExecutionContext(this._nodeSession, event.context), this._logger);
+    });
+    await this._nodeSession.send('Runtime.enable', {}).catch(e => {});
+    this._nodeElectronHandle = await this._nodeExecutionContext!.evaluateHandleInternal(() => {
+      // Resolving the race between the debugger and the boot-time script.
+      if ((global as any)._playwrightRun)
+        return (global as any)._playwrightRun();
+      return new Promise(f => (global as any)._playwrightRunCallback = f);
+    });
+  }
+
+  async evaluate<R, Arg>(pageFunction: types.FuncOn<any, Arg, R>, arg: Arg): Promise<R>;
+  async evaluate<R>(pageFunction: types.FuncOn<any, void, R>, arg?: any): Promise<R>;
+  async evaluate<R, Arg>(pageFunction: types.FuncOn<any, Arg, R>, arg: Arg): Promise<R> {
+    return this._nodeElectronHandle!.evaluate(pageFunction, arg);
+  }
+
+  async evaluateHandle<R, Arg>(pageFunction: types.FuncOn<any, Arg, R>, arg: Arg): Promise<types.SmartHandle<R>>;
+  async evaluateHandle<R>(pageFunction: types.FuncOn<any, void, R>, arg?: any): Promise<types.SmartHandle<R>>;
+  async evaluateHandle<R, Arg>(pageFunction: types.FuncOn<any, Arg, R>, arg: Arg): Promise<types.SmartHandle<R>> {
+    return this._nodeElectronHandle!.evaluateHandle(pageFunction, arg);
+  }
+
+  protected _computeDeadline(options?: types.TimeoutOptions): number {
+    return this._timeoutSettings.computeDeadline(options);
+  }
+}
+
+export class Electron  {
+  async launch(executablePath: string, options: ElectronLaunchOptions = {}): Promise<ElectronApplication> {
+    const {
+      args = [],
+      env = process.env,
+      handleSIGINT = true,
+      handleSIGTERM = true,
+      handleSIGHUP = true,
+      timeout = 30000,
+    } = options;
+    const deadline = helper.monotonicTime() + timeout;
```

Comment:
Use the `TimeoutSettings.computeDeadline` to handle zero timeout.

Reply:
Done